ui/src/components/Canvas.tsx

```diff
@@ -1090,17 +1123,150 @@ export function Canvas() {
   const [client, setClient] = useState({ x: 0, y: 0 });
 
   const onPaneContextMenu = (event) => {
+    console.log("onPaneContextMenu", event);
     event.preventDefault();
     setShowContextMenu(true);
     setPoints({ x: event.pageX, y: event.pageY });
     setClient({ x: event.clientX, y: event.clientY });
+    console.log(showContextMenu, points, client);
   };
 
+  const pasteCodePod = useCallback(
+    ({ x, y, pod }) => {
+      const reactFlowBounds = reactFlowWrapper.current.getBoundingClientRect();
+      let [posX, posY] = [
+        reactFlowBounds.width / 2,
+        reactFlowBounds.height / 2,
+      ];
+      const position = reactFlowInstance.project({ x: posX, y: posY });
+      position.x = (position.x - pod.width! / 2) as number;
+      position.y = (position.y - (pod.height ?? 0) / 2) as number;
+      console.log("pasteCodePod", position, clientId, pod);
+
+      const style = {
+        width: pod.width,
+        height: undefined,
+        opacity: 0.5,
+      };
+
+      const id = nanoid();
+      const newNode = {
+        id,
+        type: "code",
+        position,
+        data: {
+          name: pod?.name || "",
+          label: id,
+          parent: "ROOT",
+          clientId,
+        },
+        level: 0,
+        extent: "parent",
+        parentNode: undefined,
+        dragHandle: ".custom-drag-handle",
+        style,
+      };
+
+      addPod(apolloClient, {
```

Comment:
In the future, we can defer the write to DB until the final drop of the node (when we allow a user to press Esc to cancel the paste). No need to implement that because that seems to require non-trivial logic. Add a TODO comment is good enough.

Reply:
Also, since we called addPod in `observer.YMapEvent.changes`:

https://github.com/codepod-io/codepod/blob/a67b5c8d0703ca6ae1d361c5ca19a7d4f0d10bd6/ui/src/lib/nodes.tsx#L106-L111

I think all `addPod` calls in `Canvas.tsx` are duplicate, right? Should we remove them? I.e. this line (Line 1170 in this PR) and the following line (Line 901 in main):

https://github.com/codepod-io/codepod/blob/a67b5c8d0703ca6ae1d361c5ca19a7d4f0d10bd6/ui/src/components/Canvas.tsx#L901

Reply:
> addPod calls in Canvas.tsx are duplicate, right?

That depends on the behavior of [`ymap.observe`](https://docs.yjs.dev/api/shared-types/y.map#observing-changes-y.mapevent). It is triggered if "other users" make the change. Would it get triggered as well if the "local user" makes the change?

Reply:
> In the future, we can defer the write to DB until the final drop of the node 

In fact, when a user press the shortcut, the half-transparent node is immediately added to the pods array in local store, and nodesMap, but have not synced to the DB (see the `addpod` arguments, I set the `dirty` false, `doRemoteAdd` will not be executed.

When the user determines the destination by clicking, I delete the temporary transparent node in `nodesMap`, and add a node with everything identical but `opacity` and `clientId` removed, then write it formally to the DB.

`[ymap.observe](https://docs.yjs.dev/api/shared-types/y.map#observing-changes-y.mapevent).` on adding events here serves like the way in which we use `useEffect` to sync `pod.name` and `pod.position` before, only sync the data in `nodesMap` with `store.pods` on each user's own side.

To avoid the duplicated pod adding in local store, I check if the pod is already in `store.pods` here: https://github.com/li-xin-yi/codepod-1/blob/experimental%2Fpaste-on-canvas/ui/src/lib/nodes.tsx#L114

Reply:
Sorry, my bad, I noticed that I set the `dirty=true` for the temporary pod adding just now. I will fix it.